### PR TITLE
Use native Node.js functions when available

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This is the same as https://github.com/babel/babel/pull/12439 but for `fs.mkdirSync(path, { recursive: true })` and for `process.allowedNodeEnvironmentFlags`: we use the native implementation when available, and fallback to the polyfill when it's not.

This allows us to more easily prepare a Babel 8 prerelease, in the spirit of https://github.com/babel/babel/pull/12440, by simply toggling the `BABEL_8_BREAKING` env variable.

- `fs.mkdirSync(filepath, { recursive: true })` is polyfilled at build time as `parseFloat(process.versions.node) >= 10.12 ? fs.mkdirSync : require("make-dir").sync` (we can't easily feature-check because we need to polyfill the `recursive` option)
- `process.allowedNodeEnvironmentFlags` is polyfilled as `process.allowedNodeEnvironmentFlags || require("node-environment-flags")`.

Note that during local development those polyfills aren't injected since they aren't needed, but we are testing them on CI (with Node.js 6 and 8).

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12458"></a>

### CODE REVIEW

File: babel.config.js [hunk 1]
```diff
@@ -180,41 +180,102 @@ function bool(value) {
...
+          t.isObjectExpression(args[1]) &&
+          args[1].properties.length === 1 &&
+          t.isIdentifier(args[1].properties[0].key, { name: "paths" }) &&
+          t.isArrayExpression(args[1].properties[0].value) &&
+          args[1].properties[0].value.elements.length === 1
+        );
+      },
+      // require.resolve's paths option has been introduced in Node.js 8.9
+      // https://nodejs.org/api/modules.html#modules_require_resolve_request_options
+      replacement: template({ syntacticPlaceholders: true })`
```
Comment (nicolo-ribaudo) [inline babel.config.js:206]: This polyfill was already there, I only moved it.

**Comment by nicolo-ribaudo** [comment]
I opened this PR because I already had it ready, but if you don't think that the added co.plexoty in `babel.config.js` file is worth it (I'm not 100% sure) I am ok with closing this.

**Review by JLHwung — APPROVED** [review]

**Review by hzoo — APPROVED** [review]
so this doesn't remove the dep, just makes a plugin that toggles on or off the require to the dep based on node version - sounds fine since we will remove in v8 - and CI test should cover it?

**Comment by nicolo-ribaudo** [comment]
Exactly, we are testing both the behaviors: on Node.js > 10 we test the new built-in functions, while the old functions are tested on Node.js 6 and 8.

The dependency is still there (because of the polyfill), but this PR makes it so that we will only have to update `package.json` for Babel 8 and not the actual code.

**Pull request merged**